src/librustdoc/html/markdown.rs

```diff
@@ -530,6 +534,7 @@ impl<'a, 'b, 'ids, I: Iterator<Item = SpannedEvent<'a>>> Iterator
                 self.buf.push_front((Event::Html(format!("{} ", sec).into()), 0..0));
             }
 
+            let level = std::cmp::min(level + self.level + 1, MAX_HEADER_LEVEL);
```

Comment:
When `level = 0`, you get `# My Header` :arrow_right: `<h2>My Header</h2>`. This seems weird to me.

I would have expected that `HeadingLinks { level: 0 }` would mean that headings aren't getting modified. Instead, you've made it so that there is no way to tell HeadingLinks to not bump the levels. It's also silently changing how the standalone markdown files (stuff like `/src/doc/index.md`) are generated, and it's why the error index got changed.

Reply:
Completely forgot about the standalone markdown files, this is an excellent point!

Reply:
> When `level = 0`, you get `# My Header` :arrow_right: `<h2>My Header</h2>`. This seems weird to me.

I wasn't feeling great about it either but since I'm not the only thinking this, we should definitely make the heading level match the level we use.

Reply:
Just to be clear - you want doc comments in the error index to behave differently (`# this is an h1`) from doc comments in std docs (`# this is an h2`)?

Reply:
I'm not sure why it makes sense to talk about them "behaving the same" as doc comments. They're totally different.

For one thing, the biggest reason why this probably has to be fixed by changing rustdoc's behavior is that there's no way for the end-user to know what the right heading level is. The same exact doc comment might be rendered at multiple levels in multiple pages, like [slice::first on the slice page](https://doc.rust-lang.org/stable/std/primitive.slice.html#method.first) (which is a fourth-level header, so the items in the doc comment should be h5), and [slice::first on Vec](https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.first) (which is an h3, so its doc comment headers should start at h4).

For the error index, you're probably in the right here. The error index pages each have h1's at the beginning, and when they're viewed as individual documents, this works (they're `.md` files, so you expect them to be viewed using things like GitHub's markdown preview pane, or inside an IDE). But when they're combined to form the full error index, the resulting page has multiple h1's.

**But for standalone markdown files, this seems like a breaking change made with insufficient motivation.** Those aren't being combined to form huge mega-documents, and [it's a stable feature](https://doc.rust-lang.org/stable/rustdoc/what-is-rustdoc.html#using-standalone-markdown-files), not just some internal thing used by the compiler, so changing its behavior should only be done with very compelling reasons.

Reply:
The short version: yes to changing doc comments, yes to changing the error index, no to changing standalone markdown files.

Reply:
Great - this is exactly the context I needed, thanks for explaining. I agree with you. I'll update the code + maybe add a test for this standalone markdown feature if I can figure out how to do that.